### PR TITLE
z_en skb, 1 non-matching, with some documentation

### PR DESCRIPTION
**Pull request by jpburnett**
Don't know which is more dead inside, me or stalchild...

There is an interesting cast going on in func_80AFD6CC, could be real, not sure though. 

The one non-matching function is func_80AFDF24 (EnSkb_PostLimbDraw). There is a regalloc issue that I couldn't quite figure out. Who ever figures it out, I'd like to learn/know what the fix was!

### CODE REVIEW

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 1]
```diff
@@ -66,55 +74,499 @@ static ColliderJntSphInit D_80AFE068 = {
...
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Skb/func_80AFD540.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Skb/func_80AFD59C.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Skb/func_80AFD644.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Skb/func_80AFD6CC.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Skb/func_80AFD7B4.s")
+static DamageTable D_80AFE078 = {
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:78]:
```suggestion
static DamageTable sDamageTable= {
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 2]
```diff
@@ -66,55 +74,499 @@ static ColliderJntSphInit D_80AFE068 = {
...
+    func_80033480(globalCtx, &pos, 10.0f, 1, 150, 0, 1);
+}
+
+void EnSkb_Init(Actor* thisx, GlobalContext* globalCtx) {
+    EnSkb* this = THIS;
+    s16 paramOffsetBody;
+    s16 paramOffsetArm;
+
+    Actor_ProcessInitChain(&this->actor, sInitChain);
+    this->actor.colChkInfo.damageTable = &D_80AFE078;
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:163]:
```suggestion
    this->actor.colChkInfo.damageTable = &sDamageTable;
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 3]
```diff
@@ -9,25 +10,32 @@ void EnSkb_Destroy(Actor* thisx, GlobalContext* globalCtx);
...
+void func_80AFDF24(GlobalContext* globalCtx, s32 limbIndex, Gfx** dList, Vec3s* rot, void* thisx);
+
+extern SkeletonHeader D_060041F8;
+extern AnimationHeader D_06001854;
+extern AnimationHeader D_060009DC;
+extern AnimationHeader D_06000D98;
+extern AnimationHeader D_060047E0;
+extern AnimationHeader D_06000460;
 
 static ColliderJntSphElementInit D_80AFE020[2] = {
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:40]:
```suggestion
static ColliderJntSphElementInit sJntSphElementsInit[2] = {
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 4]
```diff
@@ -66,55 +74,499 @@ static ColliderJntSphInit D_80AFE068 = {
...
+    OBJECT_SKB,
+    sizeof(EnSkb),
+    (ActorFunc)EnSkb_Init,
+    (ActorFunc)EnSkb_Destroy,
+    (ActorFunc)EnSkb_Update,
+    (ActorFunc)EnSkb_Draw,
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Skb/func_80AFD968.s")
+Vec3f D_80AFE0B8 = { 0.0f, 8.0f, 0.0f };
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:125]:
```suggestion
static Vec3f sVel= { 0.0f, 8.0f, 0.0f };
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 5]
```diff
@@ -66,55 +74,499 @@ static ColliderJntSphInit D_80AFE068 = {
...
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Skb/EnSkb_Update.s")
+Vec3f D_80AFE0C4 = { 0.0f, -1.5f, 0.0f };
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:127]:
```suggestion
static Vec3f sAcell = { 0.0f, -1.5f, 0.0f };
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 6]
```diff
@@ -66,55 +74,499 @@ static ColliderJntSphInit D_80AFE068 = {
...
+    ICHAIN_F32_DIV1000(gravity, -2000, ICHAIN_STOP),
+};
 
+void EnSkb_SetupAction(EnSkb* this, EnSkbActionFunc actionFunc) {
+    this->actionFunc = actionFunc;
+}
+
+void EnSkb_SpawnDebris(GlobalContext* globalCtx, EnSkb* this, Vec3f* spawnPos) {
+    Vec3f pos;
+    Vec3f vel = D_80AFE0B8;
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:140]:
```suggestion
    Vec3f vel = sVel;
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 7]
```diff
@@ -66,55 +74,499 @@ static ColliderJntSphInit D_80AFE068 = {
...
+    Vec3f accel = D_80AFE0C4;
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:141]:
```suggestion
    Vec3f accel = sAcell;
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c (hunk 6 above)
Comment (louist103) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:140]: Can these be inlined data?

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 8]
```diff
@@ -66,55 +74,499 @@ static ColliderJntSphInit D_80AFE068 = {
...
+
+    spreadAngle = (Rand_ZeroOne() - 0.5f) * 6.28f;
+    pos.y = this->actor.floorHeight;
+    pos.x = (Math_SinF(spreadAngle) * 15.0f) + spawnPos->x;
+    pos.z = (Math_CosF(spreadAngle) * 15.0f) + spawnPos->z;
+    accel.x = Rand_CenteredFloat(1.0f);
+    accel.z = Rand_CenteredFloat(1.0f);
+    vel.y += ((Rand_ZeroOne() - 0.5f) * 4.0f);
+    scale = (Rand_ZeroOne() * 5.0f) + 12.0f;
+    EffectSsHahen_Spawn(globalCtx, &pos, &vel, &accel, 2, (scale * 0.8f), -1, 10, 0);
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:153]:
```suggestion
    EffectSsHahen_Spawn(globalCtx, &pos, &vel, &accel, 2, scale * 0.8f, -1, 10, 0);
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 9]
```diff
@@ -66,55 +74,499 @@ static ColliderJntSphInit D_80AFE068 = {
...
+    f32 spreadAngle;
+    f32 scale;
+
+    spreadAngle = (Rand_ZeroOne() - 0.5f) * 6.28f;
+    pos.y = this->actor.floorHeight;
+    pos.x = (Math_SinF(spreadAngle) * 15.0f) + spawnPos->x;
+    pos.z = (Math_CosF(spreadAngle) * 15.0f) + spawnPos->z;
+    accel.x = Rand_CenteredFloat(1.0f);
+    accel.z = Rand_CenteredFloat(1.0f);
+    vel.y += ((Rand_ZeroOne() - 0.5f) * 4.0f);
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:151]:
```suggestion
    vel.y += (Rand_ZeroOne() - 0.5f) * 4.0f;
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 10]
```diff
@@ -66,55 +74,499 @@ static ColliderJntSphInit D_80AFE068 = {
...
+    }
+    if ((this->actor.bgCheckFlags & 1) != 0) {
+        if (this->actor.speedXZ < 0.0f) {
+            this->actor.speedXZ += 0.05f;
+        }
+    }
+    if ((this->actor.colorFilterTimer == 0) && ((this->actor.bgCheckFlags & 1) != 0)) {
+        if (this->actor.colChkInfo.health == 0) {
+            func_80AFD7B4(this, globalCtx);
+            return;
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:370]: Else here and get rid of early return if possible.

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 11]
```diff
@@ -66,55 +74,499 @@ static ColliderJntSphInit D_80AFE068 = {
...
+    this->actor.flags &= ~1;
+    func_80032E24(&this->unk_28C, 18, globalCtx);
+    this->unk_283 |= 4;
+    EffectSsDeadSound_SpawnStationary(globalCtx, &this->actor.projectedPos, NA_SE_EN_STALKID_DEAD, 1, 1, 0x28);
+    EnSkb_SetupAction(this, func_80AFD880);
+}
+
+void func_80AFD880(EnSkb* this, GlobalContext* globalCtx) {
+    if (func_8003305C(&this->actor, &this->unk_28C, globalCtx, 1) != 0) {
+        if (0.01f == this->actor.scale.x) {
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:428]:
```suggestion
        if (this->actor.scale.x == 0.01f) {
```
Does the comparison need to go in that order?

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 12]
```diff
@@ -66,55 +74,499 @@ static ColliderJntSphInit D_80AFE068 = {
...
+
+    if (limbIndex == 11) {
+        if ((this->unk_283 & 2) == 0) {
+            OPEN_DISPS(globalCtx->state.gfxCtx, "../z_en_skb.c", 972);
+            color = ABS((s16)(Math_SinS((globalCtx->gameplayFrames * 0x1770)) * 95.0f)) + 0xA0;
+            gDPPipeSync(POLY_OPA_DISP++);
+            gDPSetEnvColor(POLY_OPA_DISP++, color, color, color, 0xFF);
+            CLOSE_DISPS(globalCtx->state.gfxCtx, "../z_en_skb.c", 978);
+        } else {
+            *dList = 0;
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:540]:
```suggestion
            *dList = NULL;
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 13]
```diff
@@ -66,55 +74,499 @@ static ColliderJntSphInit D_80AFE068 = {
...
+        }
+    } else if ((limbIndex == 12) && ((this->unk_283 & 2) != 0)) {
+        *dList = 0;
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:543]:
```suggestion
        *dList = NULL;
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 14]
```diff
@@ -66,55 +74,499 @@ static ColliderJntSphInit D_80AFE068 = {
...
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Skb/EnSkb_Draw.s")
+#endif
+
+void EnSkb_Draw(Actor* thisx, GlobalContext* globalCtx) {
+    EnSkb* this = THIS;
+    func_80093D18(globalCtx->state.gfxCtx);
+    SkelAnime_DrawOpa(globalCtx, this->skelAnime.skeleton, this->skelAnime.jointTable, EnSkb_OverrideLimbDraw,
+                      func_80AFDF24, &this->actor);
+}
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:572]:
```suggestion
}

```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 15]
```diff
@@ -66,55 +74,499 @@ static ColliderJntSphInit D_80AFE068 = {
...
+        this->actor.speedXZ = 0.0f;
+    }
+    Audio_PlayActorSound2(&this->actor, NA_SE_EN_GOMA_JR_FREEZE);
+    this->unk_281 = 0;
+    this->unk_280 = 6;
+    EnSkb_SetupAction(this, func_80AFD59C);
+}
+
+void func_80AFD59C(EnSkb* this, GlobalContext* globalCtx) {
+
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:358]:
```suggestion
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 16]
```diff
@@ -66,55 +74,499 @@ static ColliderJntSphInit D_80AFE068 = {
...
+    this->actor.focus.pos = this->actor.world.pos;
+    this->actor.colChkInfo.mass = 0xFE;
+    this->actor.colChkInfo.health = 2;
+    this->actor.shape.yOffset = -8000.0f;
+    SkelAnime_Init(globalCtx, &this->skelAnime, &D_060041F8, &D_06001854, this->limbDrawTable,
+                   this->transitionDrawTable, 20);
+    this->actor.naviEnemyId = 0x55;
+
+    Collider_InitJntSph(globalCtx, &this->collider);
+    Collider_SetJntSph(globalCtx, &this->collider, &this->actor, &D_80AFE068, this->colliderItem);
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:174]:
```suggestion
    Collider_SetJntSph(globalCtx, &this->collider, &this->actor, &sJntSphInit , this->colliderItem);
```
Also dont forget to rename the collider init above and make it `static`

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 17]
```diff
@@ -66,55 +74,499 @@ static ColliderJntSphInit D_80AFE068 = {
     2,
     D_80AFE020,
```
Comment (louist103) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:75]: sJntSphElementsInit

**Review by louist103 — CHANGES_REQUESTED** [review]
Looks good. Just some small changes.

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c (hunk 6 above)
Comment (jpburnett) [reply src/overlays/actors/ovl_En_Skb/z_en_skb.c:140]: Not that I know of? When I inlined the data the offset was too much:
```C
20:    addiu   t7,t7,%lo(sVelocity)               i 20:    addiu   t7,t7,%lo(.data+0xa0)
```

Comment (louist103) [reply src/overlays/actors/ovl_En_Skb/z_en_skb.c]: Did you try it with make or -mw? Also it might have to be non static.

**Review by louist103 — APPROVED** [review]

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 18]
```diff
@@ -64,57 +72,496 @@ static ColliderJntSphInit D_80AFE068 = {
...
+    Audio_PlayActorSound2(&this->actor, NA_SE_EN_RIVA_APPEAR);
+    EnSkb_SetupAction(this, func_80AFCE5C);
+}
+
+void func_80AFCE5C(EnSkb* this, GlobalContext* globalCtx) {
+    if (this->skelAnime.curFrame < 4.0f) {
+        this->actor.world.rot.y = this->actor.yawTowardsPlayer;
+        this->actor.shape.rot.y = this->actor.yawTowardsPlayer;
+    } else {
+        this->actor.flags = this->actor.flags | 1;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:223]:
```suggestion
        this->actor.flags |= 1;
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 19]
```diff
@@ -64,57 +72,496 @@ static ColliderJntSphInit D_80AFE068 = {
...
+    if ((SkelAnime_Update(&this->skelAnime) != 0) && (0.0f == this->actor.shape.yOffset)) {
+        func_80AFCD60(this);
+    }
+}
+
+void func_80AFCF48(EnSkb* this) {
+    Animation_Change(&this->skelAnime, &D_06001854, -1.0f, Animation_GetLastFrame(&D_06001854), 0.0f, 2, -4.0f);
+    this->unk_280 = 0;
+    this->unk_281 = 0;
+    this->actor.flags = this->actor.flags & ~1;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:239]:
```suggestion
    this->actor.flags &= ~1;
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 20]
```diff
@@ -64,57 +72,496 @@ static ColliderJntSphInit D_80AFE068 = {
...
+        func_80AFCF48(this);
+    } else if ((func_8002E084(&this->actor, 0x11C7) != 0) &&
+               (this->actor.xzDistToPlayer < (60.0f + (this->actor.params * 6.0f)))) {
+        func_80AFD33C(this);
+    }
+}
+
+void func_80AFD33C(EnSkb* this) {
+    Animation_Change(&this->skelAnime, &D_06000460, 0.6f, 0.0f, Animation_GetLastFrame(&D_06000460), 3, 4.0f);
+    this->collider.base.atFlags = this->collider.base.atFlags & 0xFFFB;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:305]:
```suggestion
    this->collider.base.atFlags &= ~4;
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 21]
```diff
@@ -64,57 +72,496 @@ static ColliderJntSphInit D_80AFE068 = {
...
+
+    frameData = this->skelAnime.curFrame;
+    if (frameData == 3) {
+        Audio_PlayActorSound2(&this->actor, NA_SE_EN_STALKID_ATTACK);
+        this->unk_281 = 1;
+    } else if (frameData == 6) {
+        this->unk_281 = 0;
+    }
+    if ((this->collider.base.atFlags & 4) != 0) {
+        this->collider.base.atFlags &= 0xFFF9;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:322]:
```suggestion
    if (this->collider.base.atFlags & 4) {
        this->collider.base.atFlags &= ~6;
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 22]
```diff
@@ -64,57 +72,496 @@ static ColliderJntSphInit D_80AFE068 = {
...
+        func_80AFD47C(this);
+    } else if (SkelAnime_Update(&this->skelAnime) != 0) {
+        func_80AFCD60(this);
+    }
+}
+
+void func_80AFD47C(EnSkb* this) {
+    Animation_Change(&this->skelAnime, &D_06000460, -0.4f, this->skelAnime.curFrame - 1.0f, 0.0f, 3, 0.0f);
+    this->collider.base.atFlags &= 0xFFFB;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:331]:
```suggestion
    this->collider.base.atFlags &= ~4;
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 23]
```diff
@@ -64,57 +72,496 @@ static ColliderJntSphInit D_80AFE068 = {
...
+        this->actor.speedXZ = 0.0f;
+    }
+    Audio_PlayActorSound2(&this->actor, NA_SE_EN_GOMA_JR_FREEZE);
+    this->unk_281 = 0;
+    this->unk_280 = 6;
+    EnSkb_SetupAction(this, func_80AFD59C);
+}
+
+void func_80AFD59C(EnSkb* this, GlobalContext* globalCtx) {
+    if ((this->actor.bgCheckFlags & 2) != 0) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:354]:
```suggestion
    if (this->actor.bgCheckFlags & 2) {
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 24]
```diff
@@ -64,57 +72,496 @@ static ColliderJntSphInit D_80AFE068 = {
...
+        this->actor.speedXZ = 0.0f;
+    }
+    if ((this->actor.bgCheckFlags & 1) != 0) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:357]:
```suggestion
    if (this->actor.bgCheckFlags & 1) {
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 25]
```diff
@@ -64,57 +72,496 @@ static ColliderJntSphInit D_80AFE068 = {
...
+        if (this->actor.speedXZ < 0.0f) {
+            this->actor.speedXZ += 0.05f;
+        }
+    }
+    if ((this->actor.colorFilterTimer == 0) && ((this->actor.bgCheckFlags & 1) != 0)) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:362]:
```suggestion
    if ((this->actor.colorFilterTimer == 0) && (this->actor.bgCheckFlags & 1)) {
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 26]
```diff
@@ -64,57 +72,496 @@ static ColliderJntSphInit D_80AFE068 = {
...
+    s16 phi_v1;
+    Player* player;
+
+    if ((this->unk_280 != 1) && (this->actor.bgCheckFlags & 0x60) && (this->actor.yDistToWater >= 40.0f)) {
+        this->actor.colChkInfo.health = 0;
+        this->unk_281 = 0;
+        func_80AFD7B4(this, globalCtx);
+    } else if (this->unk_280 >= 3) {
+        if ((this->collider.base.acFlags & 2) != 0) {
+            this->collider.base.acFlags &= 0xFFFD;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:453]:
```suggestion
        if (this->collider.base.acFlags & 2) {
            this->collider.base.acFlags &= ~2;
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 27]
```diff
@@ -64,57 +72,496 @@ static ColliderJntSphInit D_80AFE068 = {
...
+
+s32 EnSkb_OverrideLimbDraw(GlobalContext* globalCtx, s32 limbIndex, Gfx** dList, Vec3f* pos, Vec3s* rot, void* thisx) {
+    EnSkb* this = THIS;
+    s16 color;
+    s16 pad[2];
+
+    if (limbIndex == 11) {
+        if ((this->unk_283 & 2) == 0) {
+            OPEN_DISPS(globalCtx->state.gfxCtx, "../z_en_skb.c", 972);
+            color = ABS((s16)(Math_SinS((globalCtx->gameplayFrames * 0x1770)) * 95.0f)) + 0xA0;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:530]: we use decimal for colors so I would have `160` here
```suggestion
            color = ABS((s16)(Math_SinS((globalCtx->gameplayFrames * 0x1770)) * 95.0f)) + 160;
```

File: src/overlays/actors/ovl_En_Skb/z_en_skb.c [hunk 28]
```diff
@@ -64,57 +72,496 @@ static ColliderJntSphInit D_80AFE068 = {
...
+            gDPPipeSync(POLY_OPA_DISP++);
+            gDPSetEnvColor(POLY_OPA_DISP++, color, color, color, 0xFF);
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Skb/z_en_skb.c:532]:
```suggestion
            gDPSetEnvColor(POLY_OPA_DISP++, color, color, color, 255);
```

**Review by Roman971 — COMMENTED** [review]
Well done on this actor, just a few style things

**Review by fig02 — APPROVED** [review]

**Pull request merged**